firedrake/cofunction.py

```diff
@@ -119,11 +120,14 @@ def split(self):
 
     @utils.cached_property
     def _components(self):
-        if self.function_space().value_size == 1:
+        if self.function_space().rank == 0:
             return (self, )
         else:
-            return tuple(type(self)(self.function_space().sub(i), val=op2.DatView(self.dat, i))
-                         for i in range(self.function_space().value_size))
+            if self.dof_dset.cdim == 1:
+                return (type(self)(self.function_space().sub(0), val=self.dat),)
+            else:
+                return tuple(type(self)(self.function_space().sub(i), val=op2.DatView(self.dat, j))
+                             for i, j in enumerate(np.ndindex(self.dof_dset.dim)))
```

Comment:
What is this doing?

Reply:
The logic here was out of date, this just brings it in line with `Function._components`. 
If we have a scalar element (`rank == 0`) then `sub` just returns `self`. If we have a `VectorElement` or `TensorElement` then we return a view into the relevant component. The switch on `dof_dset.cdim` is because `DatView` complains about viewing a non-mixed `Dat`.